tchannel/container/heap.py

```diff
@@ -0,0 +1,86 @@
+from __future__ import absolute_import
+
+
+class HeapOperation(object):
+    def cmp(self, i, j):
+        """Compare the items in position i and j of the heap.
+
+        :param i: the first item's position of the heap list
+        :param j: the second item's position of the heap list
+        :return true if items[i] < items[j], otherwise false.
+        """
+        raise NotImplementedError()
+
+    def push(self, x):
+        """Push a new item into heap"""
+        raise NotImplementedError()
+
+    def pop(self):
+        """Pop an item from the heap"""
+        raise NotImplementedError()
+
+    def swap(self, i, j):
+        """swap items between position i and j of the heap"""
+        raise NotImplementedError()
+
+
+def init(h):
+    # heapify
+    n = len(h)
+    for i in range(n/2 - 1, -1):
+        down(h, i, n)
+
+
+def push(h, x):
+    h.push(x)
+    up(h, len(h)-1)
+
+
+def pop(h):
+    n = len(h) - 1
+    h.swap(0, n)
+    down(h, 0, n)
+    return h.pop()
+
+
+def remove(h, i):
+    n = len(h) - 1
+    if n != i:
+        h.swap(i, n)
+        down(h, i, n)
+        up(h, i)
+
+    return h.pop()
+
+
+def fix(h, i):
+    down(h, i, len(h))
+    up(h, i)
```

Comment:
Wait, if it turns out that you ended up sifting down, that means that you don't need to sift up, right?

Like if you have something where the heap property is already true, and then you change an item so that it is bigger than its children. `down` will swap it with the smaller of its two children. Now that new value at the index of the changed value doesn't need to be sifted up because we know that the heap property was true before, so the new value is greater than the parent of the value that changed.

Like if you have,

              i-1/2               i            i*2+1   i*2+2
    x = [..., parent_parent, ..., parent, ..., child1, child2, ...]

The heap property is already true so

    parent_parent < parent
    parent < child1
    parent < child2

which implies,

    parent_parent < child1
    parent_parent < child2

If you now change parent to violate the heap property so that it gets sifted down, that is, you change parent to be > child1 or child2, then,

              i-1/2               i            i*2+1   i*2+2
    x = [..., parent_parent, ..., child1, ..., parent, child2, ...]

Since only `parent` changed, `parent_parent < child1` is still true, so you don't need to `up(h, i)` anymore.